### ISSUE TITLE
Fix bug caused by interactions between "Popup#remove" and "closeOnClick" 

### ISSUE THREAD

**Issue by timiyay** [mapbox/mapbox-gl-js#2395]
**Codepen repro**: http://codepen.io/timiyay/pen/LNOWrr
**mapbox-gl-js version**: 0.16.0 (on `earcut-final` branch)
### Steps to Trigger Behavior
1. create an instance of `mapboxgl.Popup`, with options `{ closeButton: false, closeOnClick: true }`
2. create a click listener that calls `queryRenderedFeatures`, and displays any popups under the clicked point, or calls `popup.remove()` if no features found
3. use the previously-created instance of `mapboxgl.Popup` to render the popup
### Expected Behavior

Clicking on a feature once: should render the popup
Clicking on the same feature again: should keep the popup on the screen
Clicking on a map area with no features: should remove the popup
Clicking on a different feature: should reuse the popup, and render it on the new feature
### Actual Behavior

Clicking on a feature once: renders the popup :+1: 
Clicking on the same feature again: removes the popup :cry: 
Clicking on a map area with no features: removes the popup, or simply doesn't render a popup :+1: 
Clicking on a different feature: doesn't render any popup :cry: 

There's definitely scope for this to be programmer error, but our use is (to my eyes) identical to the Mapbox example at https://www.mapbox.com/mapbox-gl-js/example/popup-on-click/

**Comment by mourner** [comment]
Hmm, can you figure out what's the main difference between your codepen and the Mapbox example?

**Comment by timiyay** [comment]
Comparing the Codepen vs the Mapbox example...

The clear difference is that the Codepen create a single `mapboxgl.Popup()` instance, and attempts to reuse it, like so:

```js
// Create the map

// Create a single instance for our popup
var popup = new mapboxgl.Popup({
  closeButton: false,
  closeOnClick: true
});

map.on("click", function(e) {
  // get a list of features under the clicked point, using map.queryRenderedFeatures()
  
  // Note that we're reusing the popup object we created above, and calling remove()
  // on it if we've clicking a featureless area
  if (!features.length) {
    popup.remove();
    return;
  }
  
  // Again, we're reusing the same instance of mapboxgl.Popup
  popup.setLngLat(feature[0].geometry.coordinates)
    .setHTML(feature[0].properties.name)
    .addTo(map);
});

```

...whereas the Mapbox example is creating a new instance of `mapboxgl.Popup()` every time

```js
// When a click event occurs near a marker icon, open a popup at the location of
// the feature, with description HTML from its properties.
map.on('click', function (e) {
    var features = map.queryRenderedFeatures(e.point, { layers: ['markers'] });

    if (!features.length) {
        return;
    }

    var feature = features[0];

    // Populate the popup and set its coordinates
    // based on the feature found.
    var popup = new mapboxgl.Popup()
        .setLngLat(feature.geometry.coordinates)
        .setHTML(feature.properties.description)
        .addTo(map);
});
```

I may refactor to go down this route. However, my expectation is that the `mapboxgl.Popup()` instance should be reusable, as that was the subject of a previous GH issue (https://github.com/mapbox/mapbox-gl-js/issues/1477#issuecomment-139620538)

**Comment by jfirebaugh** [comment]
Change `closeOnClick: true` to `closeOnClick: false`. This code handles closing/removing the popup itself, and running both the `closeOnClick` code and the manual removal is probably triggering an invalid state in the popup.

We can probably make this work even with `closeOnClick: true`, so I'll leave the issue open to investigate that.

**Comment by timiyay** [comment]
thanks @jfirebaugh, that change does the trick. No idea why I didn't think of it trying it myself. Possibly, we should've been using `Popup()` like that all along.

Interestingly, this only became an issue for us when moving to 0.16.0. So perhaps our incorrect usage of `Popup()` was only working due to a now-patched bug?

At any rate, I consider my issue solved. Thanks.

**Comment by NealHumphrey** [comment]
After spending an hour or two hunting down why my code was triggering this error, I found this and it also solved my problem. But, I'm manually adding back the on-click behavior - but I also want the Popup to dissappear for other reasons (in my case, when the user hovers over a different point on the map). 

Would it be worthwhile handling the error so that if the Popup has already been removed for some reason, the console won't complain about not being able to remove it? That would allow someone to set closeOnClick: true but still not get an error if the remove the popup some other way.

**Comment by jfirebaugh** [comment]
Yeah, I guess there's some issue with `Popup#remove` if the popup has already been removed? Want to investigate and submit a PR?

**Comment by NealHumphrey** [comment]
Submitted PR in 5059

**Comment by jfirebaugh** [comment]
This was fixed in 65afc08ef02f5eb2081c2e3ee83f23e61722ac14 and a test is being added in #5114.
